[visionOS] LinearMediaPlayer does not display video title and artwork

aestes opened the pull request:
#### 8aeda8855e282db5d8c1b0f047a4cfb7fb0dc53d
<pre>
[visionOS] LinearMediaPlayer does not display video title and artwork
<a href="https://bugs.webkit.org/show_bug.cgi?id=272348">https://bugs.webkit.org/show_bug.cgi?id=272348</a>
<a href="https://rdar.apple.com/124234028">rdar://124234028</a>

Reviewed by Eric Carlson.

Taught PlaybackSessionInterfaceLMK to set ContentMetadata on its LinearMediaPlayer consisting of the
title, subtitle, and artwork vended to Now Playing. Since Now Playing metadata can change
arbitrarily after entering fullscreen, this requried adding a mechanism to observe changes to
Now Playing.

Added a WeakHashSet of NowPlayingMetadataObservers to PlatformMediaSessionManager and taught WebPage
to add itself as an observer when requested by its proxy. These observers are called by
MediaSessionManagerCocoa when m_nowPlayingInfo&apos;s metadata changes. To make this easier, metadata
(title, artist, album, artwork, and sourceApplicationIdentifier) was broken out into its own struct
named NowPlayingMetadata.

Also added a WeakHashSet of NowPlayingMetadataObservers to WebPageProxy. When the first observer is
added, WebPageProxy asks WebPage to add itself as an observer of PlatformMediaSessionManager and
send a message to WebPageProxy when the observer is called. PlaybackSessionInterfaceLMK adds itself
as a WebPageProxy observer when entering fullscreen. WKWebViewTesting also adds itself as an
observer to support API testing.

Added API tests.

* Source/WTF/wtf/Forward.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediasession/MediaSession.cpp:
(WebCore::MediaSession::updateNowPlayingInfo):
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::nowPlayingInfo const):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::computeNowPlayingInfo const):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/NowPlayingManager.cpp:
(WebCore::NowPlayingManager::setNowPlayingInfo):
(WebCore::NowPlayingManager::setNowPlayingInfoPrivate):
* Source/WebCore/platform/audio/NowPlayingInfo.h:
* Source/WebCore/platform/audio/NowPlayingMetadataObserver.h: Copied from Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h.
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::nowPlayingInfo const):
(WebCore::PlatformMediaSessionManager::addNowPlayingMetadataObserver):
(WebCore::PlatformMediaSessionManager::removeNowPlayingMetadataObserver):
(WebCore::PlatformMediaSessionManager::nowPlayingMetadataChanged):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
(WebCore::PlatformMediaSessionManager::nowPlayingInfo const): Deleted.
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::setNowPlayingInfo):
(WebCore::MediaSessionManagerCocoa::updateNowPlayingInfo):
* Source/WebCore/platform/audio/glib/MediaSessionGLib.cpp:
(WebCore::MediaSessionGLib::getMetadataAsGVariant):
* Source/WebCore/platform/audio/glib/MediaSessionManagerGLib.cpp:
(WebCore::MediaSessionManagerGLib::updateNowPlayingInfo):
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModel::addNowPlayingMetadataObserver):
(WebCore::PlaybackSessionModel::removeNowPlayingMetadataObserver):
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
(WebCore::PlaybackSessionInterfaceIOS::startObservingNowPlayingMetadata):
(WebCore::PlaybackSessionInterfaceIOS::stopObservingNowPlayingMetadata):
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(WebKit::nowPlayingMetadataObserver):
(WebKit::PlaybackSessionInterfaceLMK::startObservingNowPlayingMetadata):
(WebKit::PlaybackSessionInterfaceLMK::stopObservingNowPlayingMetadata):
(WebKit::artworkData):
(WebKit::PlaybackSessionInterfaceLMK::nowPlayingMetadataChanged):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::presentFullscreen):
(WebKit::VideoPresentationInterfaceLMK::dismissFullscreen):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _requestActiveNowPlayingSessionInfo:]):
(-[WKWebView _setNowPlayingMetadataObserver:]):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::addNowPlayingMetadataObserver):
(WebKit::PlaybackSessionModelContext::removeNowPlayingMetadataObserver):
(WebKit::PlaybackSessionManagerProxy::addNowPlayingMetadataObserver):
(WebKit::PlaybackSessionManagerProxy::removeNowPlayingMetadataObserver):
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::resetState):
(WebKit::WebPageProxy::addNowPlayingMetadataObserver):
(WebKit::WebPageProxy::removeNowPlayingMetadataObserver):
(WebKit::WebPageProxy::setNowPlayingMetadataObserverForTesting):
(WebKit::WebPageProxy::nowPlayingMetadataChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::nowPlayingMediaTitleAndArtist):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.peculiarEntity):
(WKSLinearMediaPlayer.contentMetadata):
(WKSLinearMediaPlayer.contentMetadataPublisher):
(SwiftOnlyData.contentMetadata): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSLinearMediaContentMetadata.contentMetadata):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):
(WebKit::WebPage::startObservingNowPlayingMetadata):
(WebKit::WebPage::stopObservingNowPlayingMetadata):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NowPlayingMetadataObserver.mm: Added.
(TestWebKitAPI::TEST(NowPlayingMetadataObserver, VideoTitle)):
(TestWebKitAPI::TEST(NowPlayingMetadataObserver, VideoTitleUpdatedByMediaSession)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/large-video-test-now-playing.html:

Canonical link: <a href="https://commits.webkit.org/277341@main">https://commits.webkit.org/277341@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4688697994f1acb53d32e459961842458b6b004b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26490 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49961 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49994 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43359 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49617 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31991 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23950 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38528 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47891 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24034 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40785 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19842 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/47174 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21474 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41948 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5354 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/40581 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43672 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42359 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51869 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46800 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22340 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18686 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45822 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23615 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44844 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10444 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24399 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54299 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23333 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11146 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
```diff
@@ -855,6 +861,26 @@ WeakPtr<PlatformMediaSession> PlatformMediaSessionManager::bestEligibleSessionFo
+
+void PlatformMediaSessionManager::nowPlayingMetadataChanged(const NowPlayingMetadata& metadata)
+{
+    m_nowPlayingMetadataObservers.forEach([&] (auto& observer) {
+        observer(metadata);
+    });
```
I worry that `m_nowPlayingMetadataObservers` could be mutated during an observer callback, if calling into the player causes it to generate and event or callback we respond to.

aestes replied:
Great point. I will copy the observers before iterating.

aestes replied:
It turns out that `WeakHashSet` safely handles being mutated during iteration via `forEach`.

aestes committed:
[visionOS] LinearMediaPlayer does not display video title and artwork
https://bugs.webkit.org/show_bug.cgi?id=272348
rdar://124234028

Reviewed by Eric Carlson.

Taught PlaybackSessionInterfaceLMK to set ContentMetadata on its LinearMediaPlayer consisting of the
title, subtitle, and artwork vended to Now Playing. Since Now Playing metadata can change
arbitrarily after entering fullscreen, this requried adding a mechanism to observe changes to
Now Playing.

Added a WeakHashSet of NowPlayingMetadataObservers to PlatformMediaSessionManager and taught WebPage
to add itself as an observer when requested by its proxy. These observers are called by
MediaSessionManagerCocoa when m_nowPlayingInfo's metadata changes. To make this easier, metadata
(title, artist, album, artwork, and sourceApplicationIdentifier) was broken out into its own struct
named NowPlayingMetadata.

Also added a WeakHashSet of NowPlayingMetadataObservers to WebPageProxy. When the first observer is
added, WebPageProxy asks WebPage to add itself as an observer of PlatformMediaSessionManager and
send a message to WebPageProxy when the observer is called. PlaybackSessionInterfaceLMK adds itself
as a WebPageProxy observer when entering fullscreen. WKWebViewTesting also adds itself as an
observer to support API testing.

Added API tests.

* Source/WTF/wtf/Forward.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediasession/MediaSession.cpp:
(WebCore::MediaSession::updateNowPlayingInfo):
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::nowPlayingInfo const):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::computeNowPlayingInfo const):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/NowPlayingManager.cpp:
(WebCore::NowPlayingManager::setNowPlayingInfo):
(WebCore::NowPlayingManager::setNowPlayingInfoPrivate):
* Source/WebCore/platform/audio/NowPlayingInfo.h:
* Source/WebCore/platform/audio/NowPlayingMetadataObserver.h: Copied from Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h.
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::nowPlayingInfo const):
(WebCore::PlatformMediaSessionManager::addNowPlayingMetadataObserver):
(WebCore::PlatformMediaSessionManager::removeNowPlayingMetadataObserver):
(WebCore::PlatformMediaSessionManager::nowPlayingMetadataChanged):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
(WebCore::PlatformMediaSessionManager::nowPlayingInfo const): Deleted.
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::setNowPlayingInfo):
(WebCore::MediaSessionManagerCocoa::updateNowPlayingInfo):
* Source/WebCore/platform/audio/glib/MediaSessionGLib.cpp:
(WebCore::MediaSessionGLib::getMetadataAsGVariant):
* Source/WebCore/platform/audio/glib/MediaSessionManagerGLib.cpp:
(WebCore::MediaSessionManagerGLib::updateNowPlayingInfo):
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModel::addNowPlayingMetadataObserver):
(WebCore::PlaybackSessionModel::removeNowPlayingMetadataObserver):
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
(WebCore::PlaybackSessionInterfaceIOS::startObservingNowPlayingMetadata):
(WebCore::PlaybackSessionInterfaceIOS::stopObservingNowPlayingMetadata):
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(WebKit::nowPlayingMetadataObserver):
(WebKit::PlaybackSessionInterfaceLMK::startObservingNowPlayingMetadata):
(WebKit::PlaybackSessionInterfaceLMK::stopObservingNowPlayingMetadata):
(WebKit::artworkData):
(WebKit::PlaybackSessionInterfaceLMK::nowPlayingMetadataChanged):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::presentFullscreen):
(WebKit::VideoPresentationInterfaceLMK::dismissFullscreen):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _requestActiveNowPlayingSessionInfo:]):
(-[WKWebView _setNowPlayingMetadataObserver:]):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::addNowPlayingMetadataObserver):
(WebKit::PlaybackSessionModelContext::removeNowPlayingMetadataObserver):
(WebKit::PlaybackSessionManagerProxy::addNowPlayingMetadataObserver):
(WebKit::PlaybackSessionManagerProxy::removeNowPlayingMetadataObserver):
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::resetState):
(WebKit::WebPageProxy::addNowPlayingMetadataObserver):
(WebKit::WebPageProxy::removeNowPlayingMetadataObserver):
(WebKit::WebPageProxy::setNowPlayingMetadataObserverForTesting):
(WebKit::WebPageProxy::nowPlayingMetadataChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::nowPlayingMediaTitleAndArtist):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.peculiarEntity):
(WKSLinearMediaPlayer.contentMetadata):
(WKSLinearMediaPlayer.contentMetadataPublisher):
(SwiftOnlyData.contentMetadata): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSLinearMediaContentMetadata.contentMetadata):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):
(WebKit::WebPage::startObservingNowPlayingMetadata):
(WebKit::WebPage::stopObservingNowPlayingMetadata):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NowPlayingMetadataObserver.mm: Added.
(TestWebKitAPI::TEST(NowPlayingMetadataObserver, VideoTitle)):
(TestWebKitAPI::TEST(NowPlayingMetadataObserver, VideoTitleUpdatedByMediaSession)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/large-video-test-now-playing.html:

Canonical link: https://commits.webkit.org/277341@main
